### PR TITLE
variable: Add tests for sysvar registration, int/enum validation

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Problem Summary:

This adds unit tests for sysvar registration, int/enum validation. Most of this is already tested by integration tests.

Because validation and set functions are moved into the sysvar struct, it actually makes testing easier. In future some of the set or integration tests could probably be moved to sysvar tests.

### What is changed and how it works?

What's Changed:

Tests only.

### Related changes

- None

### Check List 

Tests 

- Unit test

Side effects

- None

### Release note 

- No release note

### CODE REVIEW

File: sessionctx/variable/sysvar_test.go [hunk 1]
```diff
@@ -87,3 +88,83 @@ func (*testSysVarSuite) TestError(c *C) {
...
+	c.Assert(val, Equals, "300")
+
+	// out of range but permitted due to auto value
+	val, err = sv.Validate(vars, "-1", ScopeSession)
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "-1")
+}
+
+func (*testSysVarSuite) TestEnumValidation(c *C) {
+	sv := SysVar{Scope: ScopeGlobal | ScopeSession, Name: "mynewsysvar", Value: "123", Type: TypeEnum, PossibleValues: []string{"OFF", "ON", "AUTO"}}
```
Comment (xiongjiwei) [inline sessionctx/variable/sysvar_test.go:148]: "123" can not be the default value

File: sessionctx/variable/sysvar_test.go [hunk 2]
```diff
@@ -87,3 +88,83 @@ func (*testSysVarSuite) TestError(c *C) {
...
+	c.Assert(err.Error(), Equals, "[variable:1231]Variable 'mynewsysvar' can't be set to the value of '301'")
+
+	_, err = sv.Validate(vars, "5", ScopeSession)
+	c.Assert(err.Error(), Equals, "[variable:1231]Variable 'mynewsysvar' can't be set to the value of '5'")
+
+	val, err := sv.Validate(vars, "300", ScopeSession)
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "300")
+
+	// out of range but permitted due to auto value
```
Comment (xiongjiwei) [inline sessionctx/variable/sysvar_test.go:141]: just a question, what happens if `-1` is a valid value and `AllowAutoValue` is true?

Comment (xhebox) [reply sessionctx/variable/sysvar_test.go:141]:
```
func (sv *SysVar) checkInt64SystemVar(value string, vars *SessionVars) (string, error) {
	if sv.AllowAutoValue && value == "-1" {
		return value, nil
	}
```
Looks like it doesn't matter. `-1` will be returned if it is valid or `AllowAutoValue == true`. But you can not use `-1` to identify "auto" anymore.

File: sessionctx/variable/sysvar_test.go (hunk 1 above)
Comment (morgo) [reply sessionctx/variable/sysvar_test.go:148]: You're right, I'll fix it.

File: sessionctx/variable/sysvar_test.go (hunk 2 above)
Comment (morgo) [reply sessionctx/variable/sysvar_test.go:141]: Yeah, if either is supported the value is permitted.

I think in most cases that use an AutoValue, it refers to something like worker threads, so it is likely always a positive value.

But the autovalue when assigned to the sysvar means nothing by itself. It is whatever reads the value that needs to also support the idea of an auto value.

Comment (morgo) [reply sessionctx/variable/sysvar_test.go:141]: AutoValue is a MySQL concept btw. It is used for a lot of settings in performance_schema for example. We are just copying the behavior for some of our own settings.

**Review by xiongjiwei — APPROVED** [review]
/lgtm

**Pull request merged**